import libraries

In [1]:
import pandas as pd
import numpy as np

### Loading the kaggle datasets

In [2]:
# load the train dataset
train = pd.read_parquet("Training.parquet")
train.head()

url  length_url  \
0  https://www.todayshomeowner.com/how-to-make-ho...          82   
1  http://thapthan.ac.th/information/confirmation...          93   
2  http://app.dialoginsight.com/T/OFC4/L2S/3888/B...         121   
3                           https://www.bedslide.com          24   
4  https://tabs.ultimate-guitar.com/s/sex_pistols...          73   

   length_hostname  ip  nb_dots  nb_hyphens  nb_at  nb_qm  nb_and  nb_or  ...  \
0               23   0        2           7      0      0       0      0  ...   
1               14   1        2           0      0      0       0      0  ...   
2               21   1        3           0      0      0       0      0  ...   
3               16   0        2           0      0      0       0      0  ...   
4               24   0        3           1      0      0       0      0  ...   

   domain_in_title  domain_with_copyright  whois_registered_domain  \
0                1                      1                        0   
1                1                      0                        1   
2                1                      1                        0   
3                0                      0                        0   
4                0                      0                        0   

   domain_registration_length  domain_age  web_traffic  dns_record  \
0                         240        8892        67860           0   
1                           0        2996      4189860           0   
2                          30        2527       346022           0   
3                         139        7531      1059151           0   
4                        3002        7590          635           0   

   google_index  page_rank      status  
0             1          4  legitimate  
1             1          2    phishing  
2             1          3    phishing  
3             0          4  legitimate  
4             1          5  legitimate  

[5 rows x 89 columns]

### load the test dataset

In [3]:
test = pd.read_parquet("Testing.parquet")
test.head()

url  length_url  \
0               https://clubedemilhagem.com/home.php          36   
1  http://www.medicalnewstoday.com/articles/18893...          51   
2     https://en.wikipedia.org/wiki/NBC_Nightly_News          46   
3  http://secure.web894.com/customer_center/custo...         185   
4  https://en.wikipedia.org/wiki/Transaction_proc...          52   

   length_hostname  ip  nb_dots  nb_hyphens  nb_at  nb_qm  nb_and  nb_or  ...  \
0               19   0        2           0      0      0       0      0  ...   
1               24   0        3           0      0      0       0      0  ...   
2               16   0        2           0      0      0       0      0  ...   
3               17   1        2           1      0      1       2      0  ...   
4               16   0        2           0      0      0       0      0  ...   

   domain_in_title  domain_with_copyright  whois_registered_domain  \
0                1                      0                        0   
1                1                      1                        0   
2                0                      1                        0   
3                1                      1                        0   
4                0                      1                        0   

   domain_registration_length  domain_age  web_traffic  dns_record  \
0                         344          21            0           0   
1                         103        6106          737           0   
2                         901        7134           12           0   
3                         247        1944            0           0   
4                         901        7134           12           0   

   google_index  page_rank      status  
0             1          0    phishing  
1             1          6  legitimate  
2             0          7  legitimate  
3             1          0    phishing  
4             0          7  legitimate  

[5 rows x 89 columns]

### Combine the train and the test dataset

In [4]:
Dataset1 = pd.concat([train, test], axis=0, ignore_index=True)
Dataset1.shape

(11430, 89)

#### Drop columns whose value counts is only 1

In [5]:
drop_columns = []
for i in range(89):
    if len(Dataset1[Dataset1.columns[i]].value_counts()) == 1:
        drop_columns.append(Dataset1.columns[i])

Dataset1.drop(columns=drop_columns,inplace = True)

Check the shape of the shape of the new dataset

In [6]:
Dataset1.shape

(11430, 83)

The new dataset has 11430 observations and 83 columns

#### Checking the columns of the dataset 1

In [7]:
Dataset1.columns

Index(['url', 'length_url', 'length_hostname', 'ip', 'nb_dots', 'nb_hyphens',
       'nb_at', 'nb_qm', 'nb_and', 'nb_eq', 'nb_underscore', 'nb_tilde',
       'nb_percent', 'nb_slash', 'nb_star', 'nb_colon', 'nb_comma',
       'nb_semicolumn', 'nb_dollar', 'nb_space', 'nb_www', 'nb_com',
       'nb_dslash', 'http_in_path', 'https_token', 'ratio_digits_url',
       'ratio_digits_host', 'punycode', 'port', 'tld_in_path',
       'tld_in_subdomain', 'abnormal_subdomain', 'nb_subdomains',
       'prefix_suffix', 'random_domain', 'shortening_service',
       'path_extension', 'nb_redirection', 'nb_external_redirection',
       'length_words_raw', 'char_repeat', 'shortest_words_raw',
       'shortest_word_host', 'shortest_word_path', 'longest_words_raw',
       'longest_word_host', 'longest_word_path', 'avg_words_raw',
       'avg_word_host', 'avg_word_path', 'phish_hints', 'domain_in_brand',
       'brand_in_subdomain', 'brand_in_path', 'suspecious_tld',
       'statistical_report', 'nb_hyper

check dataset info

In [8]:
Dataset1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11430 entries, 0 to 11429
Data columns (total 83 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   url                         11430 non-null  object 
 1   length_url                  11430 non-null  int64  
 2   length_hostname             11430 non-null  int64  
 3   ip                          11430 non-null  int64  
 4   nb_dots                     11430 non-null  int64  
 5   nb_hyphens                  11430 non-null  int64  
 6   nb_at                       11430 non-null  int64  
 7   nb_qm                       11430 non-null  int64  
 8   nb_and                      11430 non-null  int64  
 9   nb_eq                       11430 non-null  int64  
 10  nb_underscore               11430 non-null  int64  
 11  nb_tilde                    11430 non-null  int64  
 12  nb_percent                  11430 non-null  int64  
 13  nb_slash                    114

check values of the status column

In [9]:
Dataset1.status.value_counts()

phishing      5715
legitimate    5715
Name: status, dtype: int64

- The dataset has 5715 legitimate urls and 5715 illegitimate urls.
- This is a balanced dataset

In [10]:
Dataset1.head()

url  length_url  \
0  https://www.todayshomeowner.com/how-to-make-ho...          82   
1  http://thapthan.ac.th/information/confirmation...          93   
2  http://app.dialoginsight.com/T/OFC4/L2S/3888/B...         121   
3                           https://www.bedslide.com          24   
4  https://tabs.ultimate-guitar.com/s/sex_pistols...          73   

   length_hostname  ip  nb_dots  nb_hyphens  nb_at  nb_qm  nb_and  nb_eq  ...  \
0               23   0        2           7      0      0       0      0  ...   
1               14   1        2           0      0      0       0      0  ...   
2               21   1        3           0      0      0       0      0  ...   
3               16   0        2           0      0      0       0      0  ...   
4               24   0        3           1      0      0       0      0  ...   

   domain_in_title  domain_with_copyright  whois_registered_domain  \
0                1                      1                        0   
1                1                      0                        1   
2                1                      1                        0   
3                0                      0                        0   
4                0                      0                        0   

   domain_registration_length  domain_age  web_traffic  dns_record  \
0                         240        8892        67860           0   
1                           0        2996      4189860           0   
2                          30        2527       346022           0   
3                         139        7531      1059151           0   
4                        3002        7590          635           0   

   google_index  page_rank      status  
0             1          4  legitimate  
1             1          2    phishing  
2             1          3    phishing  
3             0          4  legitimate  
4             1          5  legitimate  

[5 rows x 83 columns]

#### Convert the target column to numerical

In [11]:
# Convert 'status' column to numerical
Dataset1['status'] = Dataset1['status'].map({'legitimate': 0, 'phishing': 1})
Dataset1['status'].value_counts()

1    5715
0    5715
Name: status, dtype: int64

In [12]:
Dataset1.head()

url  length_url  \
0  https://www.todayshomeowner.com/how-to-make-ho...          82   
1  http://thapthan.ac.th/information/confirmation...          93   
2  http://app.dialoginsight.com/T/OFC4/L2S/3888/B...         121   
3                           https://www.bedslide.com          24   
4  https://tabs.ultimate-guitar.com/s/sex_pistols...          73   

   length_hostname  ip  nb_dots  nb_hyphens  nb_at  nb_qm  nb_and  nb_eq  ...  \
0               23   0        2           7      0      0       0      0  ...   
1               14   1        2           0      0      0       0      0  ...   
2               21   1        3           0      0      0       0      0  ...   
3               16   0        2           0      0      0       0      0  ...   
4               24   0        3           1      0      0       0      0  ...   

   domain_in_title  domain_with_copyright  whois_registered_domain  \
0                1                      1                        0   
1                1                      0                        1   
2                1                      1                        0   
3                0                      0                        0   
4                0                      0                        0   

   domain_registration_length  domain_age  web_traffic  dns_record  \
0                         240        8892        67860           0   
1                           0        2996      4189860           0   
2                          30        2527       346022           0   
3                         139        7531      1059151           0   
4                        3002        7590          635           0   

   google_index  page_rank  status  
0             1          4       0  
1             1          2       1  
2             1          3       1  
3             0          4       0  
4             1          5       0  

[5 rows x 83 columns]

## Exploring the UCI Machine Learning Repository dataset

#### Loading UCI Dataset

In [13]:
Dataset2 = pd.read_csv("PhiUSIIL_Phishing_URL_Dataset.csv")
Dataset2.head()

FILENAME                                 URL  URLLength  \
0  521848.txt    https://www.southbankmosaics.com         31   
1   31372.txt            https://www.uni-mainz.de         23   
2  597387.txt      https://www.voicefmradio.co.uk         29   
3  554095.txt         https://www.sfnmjournal.com         26   
4  151578.txt  https://www.rewildingargentina.org         33   

                       Domain  DomainLength  IsDomainIP  TLD  \
0    www.southbankmosaics.com            24           0  com   
1            www.uni-mainz.de            16           0   de   
2      www.voicefmradio.co.uk            22           0   uk   
3         www.sfnmjournal.com            19           0  com   
4  www.rewildingargentina.org            26           0  org   

   URLSimilarityIndex  CharContinuationRate  TLDLegitimateProb  ...  Pay  \
0               100.0              1.000000           0.522907  ...    0   
1               100.0              0.666667           0.032650  ...    0   
2               100.0              0.866667           0.028555  ...    0   
3               100.0              1.000000           0.522907  ...    1   
4               100.0              1.000000           0.079963  ...    1   

   Crypto  HasCopyrightInfo  NoOfImage  NoOfCSS  NoOfJS  NoOfSelfRef  \
0       0                 1         34       20      28          119   
1       0                 1         50        9       8           39   
2       0                 1         10        2       7           42   
3       1                 1          3       27      15           22   
4       0                 1        244       15      34           72   

   NoOfEmptyRef  NoOfExternalRef  label  
0             0              124      1  
1             0              217      1  
2             2                5      1  
3             1               31      1  
4             1               85      1  

[5 rows x 56 columns]

#### check the shape of Dataset2

In [14]:
Dataset2.shape

(235795, 56)

In [23]:
Dataset2.label.value_counts()

1    134850
0    100945
Name: label, dtype: int64

The dataset has 235,975 observations with 56 columns

#### Drop columns whose value counts is only 1
FILENAME column is also dropped since it doesn't add any significance to the dataset

In [15]:
drop_columns = []
for i in range(56):
    if len(Dataset2[Dataset2.columns[i]].value_counts()) == 1:
        drop_columns.append(Dataset2.columns[i])

drop_columns.append('FILENAME')

Dataset2.drop(columns=drop_columns, inplace=True)

In [16]:
Dataset2.shape

(235795, 55)

The new dataset has 55 columns and 235,795 observations

Check the columns of dataset2

In [17]:
Dataset2.columns

Index(['URL', 'URLLength', 'Domain', 'DomainLength', 'IsDomainIP', 'TLD',
       'URLSimilarityIndex', 'CharContinuationRate', 'TLDLegitimateProb',
       'URLCharProb', 'TLDLength', 'NoOfSubDomain', 'HasObfuscation',
       'NoOfObfuscatedChar', 'ObfuscationRatio', 'NoOfLettersInURL',
       'LetterRatioInURL', 'NoOfDegitsInURL', 'DegitRatioInURL',
       'NoOfEqualsInURL', 'NoOfQMarkInURL', 'NoOfAmpersandInURL',
       'NoOfOtherSpecialCharsInURL', 'SpacialCharRatioInURL', 'IsHTTPS',
       'LineOfCode', 'LargestLineLength', 'HasTitle', 'Title',
       'DomainTitleMatchScore', 'URLTitleMatchScore', 'HasFavicon', 'Robots',
       'IsResponsive', 'NoOfURLRedirect', 'NoOfSelfRedirect', 'HasDescription',
       'NoOfPopup', 'NoOfiFrame', 'HasExternalFormSubmit', 'HasSocialNet',
       'HasSubmitButton', 'HasHiddenFields', 'HasPasswordField', 'Bank', 'Pay',
       'Crypto', 'HasCopyrightInfo', 'NoOfImage', 'NoOfCSS', 'NoOfJS',
       'NoOfSelfRef', 'NoOfEmptyRef', 'NoOfExternalRef', 'labe

### Working with both datasets

- Dataset1 and Dataset2 are both phishing datasets. We therefore need to identify similar columns as well as unique columns before combining the two datasets

## **`Common or Similar Columns`**
### **`URL Details`**

- url (Dataset 1) ↔️ URL (Dataset 2)
- length_url (Dataset 1) ↔️ URLLength (Dataset 2)
- length_hostname (Dataset 1) ↔️ DomainLength (Dataset 2)
- ip (Dataset 1) ↔️ IsDomainIP (Dataset 2) – boolean indicators of IP addresses.

### **`Special Characters`**

- Dataset 1 has detailed counts of various characters (e.g., nb_qm, nb_and, etc.) which correspond to general character counts in Dataset 2:

- nb_qm ↔️ NoOfQMarkInURL
- nb_and ↔️ NoOfAmpersandInURL
- nb_eq ↔️ NoOfEqualsInURL
- nb_dots ↔️ part of general NoOfOtherSpecialCharsInURL
- ratio_digits_url (Dataset 1) ↔️ DegitRatioInURL (Dataset 2)
- ratio_digits_host (Dataset 1) ↔️ LetterRatioInURL (Dataset 2)

### **`Subdomains`**
- nb_subdomains (Dataset 1) ↔️ NoOfSubDomain (Dataset 2)

**`HTTPS and Security`**
- https_token (Dataset 1) ↔️ IsHTTPS (Dataset 2)

### **`Domain and Brand Matches`**
- `domain_in_brand`, `brand_in_subdomain`, `brand_in_path` (Dataset 1) ↔️ `DomainTitleMatchScore` and `URLTitleMatchScore` (Dataset 2) – both reflect domain and brand associations.

**`HTML Elements & Embedded Content`**
- external_favicon (Dataset 1) ↔️ HasFavicon (Dataset 2)
- links_in_tags (Dataset 1) ↔️ NoOfExternalRef and NoOfSelfRef (Dataset 2)
- submit_email (Dataset 1) ↔️ HasSubmitButton (Dataset 2)
-iframe (Dataset 1) ↔️ NoOfiFrame (Dataset 2)
- popup_window (Dataset 1) ↔️ NoOfPopup (Dataset 2)
- domain_with_copyright (Dataset 1) ↔️ HasCopyrightInfo (Dataset 2)

### **`Miscellaneous Indicators`**
- web_traffic (Dataset 1) ↔️ Similar to Robots and IsResponsive in Dataset 2 – these columns may represent engagement or activity level indicators.

### **`Unique Columns in Each Dataset`**

#### **Dataset 1 Specific Columns:**

- Many columns are detailed counts of various special characters (e.g., `nb_at`, `nb_hyphens`, `nb_tilde`, etc.)
- Additional technical indicators like `port`, `tld_in_path`, and `statistical_report`
- `Whois` and registration details like `domain_age` and `domain_registration_length`

#### **Dataset 2 Specific Columns:**
- Contains unique columns related to the web structure, like `NoOfImage`, `NoOfCSS`, `NoOfJS`, `Bank`, `Pay`, `Crypto`, and `HasSocialNet`

#### convert the status column in Dataset 1 to numerical values

## MERGING DATASET 1 AND DATASET 2

The datasets will be merged as follows:
1. Standardize Column Names: Rename similar columns in each dataset to a common name.
2. Align Columns: Add missing columns in each dataset as NaN.
3. Concatenate: Use pd.concat to combine the datasets.

In [18]:
# Standardize similar column names
Dataset1.rename(columns={
    'url': 'URL', 'length_url': 'URLLength', 'length_hostname': 'DomainLength',
    'ip': 'IsDomainIP', 'nb_qm': 'NoOfQMarkInURL', 'nb_and': 'NoOfAmpersandInURL',
    'nb_eq': 'NoOfEqualsInURL', 'nb_subdomains': 'NoOfSubDomain',
    'https_token': 'IsHTTPS', 'external_favicon': 'HasFavicon', 
    'submit_email': 'HasSubmitButton', 'iframe': 'NoOfiFrame', 
    'popup_window': 'NoOfPopup', 'domain_with_copyright': 'HasCopyrightInfo', 'status': 'label'
    # Add other similar mappings as needed
}, inplace=True)

# Add missing columns to each dataset
for column in set(Dataset1.columns).difference(Dataset2.columns):
    Dataset2[column] = pd.NA
for column in set(Dataset2.columns).difference(Dataset1.columns):
    Dataset1[column] = pd.NA

# Combine the datasets
combined_data = pd.concat([Dataset1, Dataset2], ignore_index=True)



In [19]:
combined_data

URL  URLLength  \
0       https://www.todayshomeowner.com/how-to-make-ho...         82   
1       http://thapthan.ac.th/information/confirmation...         93   
2       http://app.dialoginsight.com/T/OFC4/L2S/3888/B...        121   
3                                https://www.bedslide.com         24   
4       https://tabs.ultimate-guitar.com/s/sex_pistols...         73   
...                                                   ...        ...   
247220                     https://www.skincareliving.com         29   
247221                      https://www.winchester.gov.uk         28   
247222                    https://www.nononsensedesign.be         30   
247223  https://patient-cell-40f5.updatedlogmylogin.wo...         55   
247224                 https://www.alternativefinland.com         33   

        DomainLength  IsDomainIP nb_dots nb_hyphens nb_at  NoOfQMarkInURL  \
0                 23           0       2          7     0               0   
1                 14           1       2          0     0               0   
2                 21           1       3          0     0               0   
3                 16           0       2          0     0               0   
4                 24           0       3          1     0               0   
...              ...         ...     ...        ...   ...             ...   
247220            22           0    <NA>       <NA>  <NA>               0   
247221            21           0    <NA>       <NA>  <NA>               0   
247222            23           0    <NA>       <NA>  <NA>               0   
247223            47           0    <NA>       <NA>  <NA>               0   
247224            26           0    <NA>       <NA>  <NA>               0   

        NoOfAmpersandInURL  NoOfEqualsInURL  ... URLCharProb HasPasswordField  \
0                        0                0  ...        <NA>             <NA>   
1                        0                0  ...        <NA>             <NA>   
2                        0                0  ...        <NA>             <NA>   
3                        0                0  ...        <NA>             <NA>   
4                        0                0  ...        <NA>             <NA>   
...                    ...              ...  ...         ...              ...   
247220                   0                0  ...   0.0587386                0   
247221                   0                0  ...   0.0538339                0   
247222                   0                0  ...   0.0630935                0   
247223                   0                0  ...   0.0502107                1   
247224                   0                0  ...   0.0605959                0   

       NoOfURLRedirect NoOfSelfRedirect LargestLineLength LetterRatioInURL  \
0                 <NA>             <NA>              <NA>             <NA>   
1                 <NA>             <NA>              <NA>             <NA>   
2                 <NA>             <NA>              <NA>             <NA>   
3                 <NA>             <NA>              <NA>             <NA>   
4                 <NA>             <NA>              <NA>             <NA>   
...                ...              ...               ...              ...   
247220               0                0              9381            0.552   
247221               0                0               348              0.5   
247222               0                0             13277            0.567   
247223               0                0              1807            0.709   
247224               0                0              3346            0.606   

       NoOfImage   Pay NoOfCSS HasTitle  
0           <NA>  <NA>    <NA>     <NA>  
1           <NA>  <NA>    <NA>     <NA>  
2           <NA>  <NA>    <NA>     <NA>  
3           <NA>  <NA>    <NA>     <NA>  
4           <NA>  <NA>    <NA>     <NA>  
...          ...   ...     ...      ...  
247220        51     1       7        1  
247221        50  

In [20]:
combined_data.label.value_counts()

1    140565
0    106660
Name: label, dtype: int64

In [21]:

# Save combined data
combined_data.to_csv('combined_phishing_data.csv', index=False)
print("Combined data saved as 'combined_phishing_data.csv'")

Combined data saved as 'combined_phishing_data.csv'
